From cc7e497f0b495fa2ca02817d8fa43e4fb7f03dd9 Mon Sep 17 00:00:00 2001
From: Sebastien PASDELOUP <sebastien.pasdeloup-ext@st.com>
Date: Mon, 22 Apr 2024 11:31:48 +0200
Subject: [PATCH 582/841] feat(stm32mp2): add TAMPER management logic at bl2
 platform setup

TAMPER occurring in deep low-power modes or power off state need to be
caught and handled in TF-A. Otherwise, the secrets can be locked and the
platform cannot boot. Add a special case for the LSE monitoring tamper
that requires a backup domain reset.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git bc4c775d377e27a31391181c1af67ccd9b9736b8 feat(stm32mp2): add TAMPER management logic at bl2 platform setup]
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: I745f6d9ad91c480d4360662862fffd2a9e90b13d
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/371239
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/373932
Tested-by: Sebastien PASDELOUP <sebastien.pasdeloup-ext@st.com>
Reviewed-by: Sebastien PASDELOUP <sebastien.pasdeloup-ext@st.com>
Domain-Review: Sebastien PASDELOUP <sebastien.pasdeloup-ext@st.com>
---
 plat/st/stm32mp2/bl2_plat_setup.c | 70 ++++++++++++++++++++++++++++---
 plat/st/stm32mp2/stm32mp2_def.h   | 11 ++++-
 2 files changed, 73 insertions(+), 8 deletions(-)

diff --git a/plat/st/stm32mp2/bl2_plat_setup.c b/plat/st/stm32mp2/bl2_plat_setup.c
index 537e10daf..cd66b2dda 100644
--- a/plat/st/stm32mp2/bl2_plat_setup.c
+++ b/plat/st/stm32mp2/bl2_plat_setup.c
@@ -222,6 +222,35 @@ void bl2_platform_setup(void)
 #endif
 }
 
+static void handle_potential_tamper(uint32_t bit_off)
+{
+	/* Fixme: Add implementation specific logic here */
+	ERROR("Handling Potential tamper\n");
+	mmio_setbits_32(TAMP_BASE + TAMP_SCR, BIT_32(bit_off));
+}
+
+static void handle_confirmed_tamper(uint32_t bit_off __unused)
+{
+	/* Fixme: Add implementation specific logic here */
+	ERROR("Handling Confirmed tamper\n");
+	panic();
+}
+
+static bool lse_tamper_detection(void)
+{
+	if ((mmio_read_32(TAMP_BASE + TAMP_SR) & TAMP_SR_LSE_MONITORING) != 0U) {
+		mmio_clrbits_32(RCC_BASE + RCC_BDCR, RCC_BDCR_LSECSSON);
+		mmio_clrbits_32(RCC_BASE + RCC_BDCR, RCC_BDCR_LSEON);
+		if (mmio_read_32(RCC_BASE + RCC_C1CIFCLRR) & RCC_C1CIFCLRR_LSECSSF) {
+			mmio_clrbits_32(RCC_BASE + RCC_C1CIFCLRR, RCC_C1CIFCLRR_LSECSSF);
+		}
+
+		return true;
+	}
+
+	return false;
+}
+
 static void reset_backup_domain(void)
 {
 #if !STM32MP_M33_TDCID
@@ -247,7 +276,7 @@ static void reset_backup_domain(void)
 	}
 #endif /* STM32MP21 */
 
-	/* Reset backup domain on cold boot cases */
+	/* Reset backup domain on cold boot cases or when LSE tamper occurred */
 	if ((mmio_read_32(rcc_base + RCC_BDCR) & RCC_BDCR_RTCCKEN) == 0U) {
 		mmio_setbits_32(rcc_base + RCC_BDCR, RCC_BDCR_VSWRST);
 
@@ -260,21 +289,47 @@ static void reset_backup_domain(void)
 #endif
 }
 
-static void stm32_tamp_check_tamper_event(void)
+static void check_tamper_event(bool lse_tamper_occured)
 {
-	uint32_t sr = mmio_read_32(TAMP_SR);
+	uint32_t sr = mmio_read_32(TAMP_BASE + TAMP_SR);
+
+	if (sr == 0U) {
+		return;
+	}
 
-	if (sr != 0U) {
+	ERROR("\n");
+	if (lse_tamper_occured) {
+		ERROR("** INTRUSION ALERT: LSE MONITORING TAMPER DETECTED **\n");
 		ERROR("\n");
+
+#if !STM32MP_M33_TDCID
+		/*
+		 * Fixme: Add logic to handle the LSE tamper here (e.g change RTC clock source
+		 * instead). This part is implementation specific.
+		 */
+		mmio_clrbits_32(RCC_BASE + RCC_BDCR, RCC_BDCR_RTCCKEN);
+		ERROR("** Rebooting... **\n");
+		stm32mp_system_reset();
+#endif
+	} else {
 		while (sr != 0U) {
 			unsigned int bit_off = __builtin_ctz(sr);
 			bool is_internal = bit_off >= TAMP_SR_INT_SHIFT;
+			uint32_t cr2 = mmio_read_32(TAMP_BASE + TAMP_CR2);
+			uint32_t cr3 = mmio_read_32(TAMP_BASE + TAMP_CR3);
 
 			ERROR("** INTRUSION ALERT: %s TAMPER %u DETECTED **\n",
 			      is_internal ? "INTERNAL" : "EXTERNAL",
 			      is_internal ? (bit_off - TAMP_SR_INT_SHIFT + 1U) : (bit_off + 1U));
 
-			sr &= ~BIT(bit_off);
+			if ((is_internal && ((cr3 & BIT_32(bit_off - TAMP_SR_INT_SHIFT)) != 0U)) ||
+			    (!is_internal && ((cr2 & BIT_32(bit_off)) != 0U))) {
+				handle_potential_tamper(bit_off);
+			} else {
+				handle_confirmed_tamper(bit_off);
+			}
+
+			sr &= ~BIT_32(bit_off);
 		}
 		ERROR("\n");
 	}
@@ -289,6 +344,7 @@ void bl2_el3_plat_arch_setup(void)
 				(boot_context->boot_interface_selected ==
 				 BOOT_API_CTX_BOOT_INTERFACE_SEL_SERIAL_UART);
 	uintptr_t uart_prog_addr __unused;
+	bool lse_tamper_occured = false;
 
 	if (stm32_otp_probe() != 0) {
 		panic();
@@ -309,6 +365,8 @@ void bl2_el3_plat_arch_setup(void)
 		panic();
 	}
 
+	lse_tamper_occured = lse_tamper_detection();
+
 	reset_backup_domain();
 
 #if !STM32MP_M33_TDCID
@@ -371,7 +429,7 @@ void bl2_el3_plat_arch_setup(void)
 	}
 
 skip_console_init:
-	stm32_tamp_check_tamper_event();
+	check_tamper_event(lse_tamper_occured);
 
 #if !TRUSTED_BOARD_BOOT
 	if (stm32mp_check_closed_device() == STM32MP_CHIP_SEC_CLOSED) {
diff --git a/plat/st/stm32mp2/stm32mp2_def.h b/plat/st/stm32mp2/stm32mp2_def.h
index 6b2c8f94d..216f53301 100644
--- a/plat/st/stm32mp2/stm32mp2_def.h
+++ b/plat/st/stm32mp2/stm32mp2_def.h
@@ -532,13 +532,20 @@ enum ddr_type {
 #define PLAT_MAX_TAMP_INT		U(5)
 #define PLAT_MAX_TAMP_EXT		U(3)
 #define TAMP_BASE			U(0x46010000)
-#define TAMP_SMCR			(TAMP_BASE + U(0x20))
-#define TAMP_SR				(TAMP_BASE + U(0x30))
+#define TAMP_CR2			U(0x4)
+#define TAMP_CR2_MASK_POM		GENMASK_32(7, 0)
+#define TAMP_CR3			U(0x8)
+#define TAMP_CR3_MASK_POM		GENMASK_32(14, 0)
+#define TAMP_SMCR			U(0x20)
+#define TAMP_SR				U(0x30)
 #define TAMP_SR_INT_SHIFT		U(16)
 #define TAMP_BKP_REG_CLK		CK_BUS_RTC
 #define TAMP_BKP_SEC_NUMBER		U(10)
+#define TAMP_SCR			U(0x3C)
 #define TAMP_COUNTR			U(0x40)
 
+#define TAMP_SR_LSE_MONITORING		BIT(18)
+
 /*******************************************************************************
  * STM32MP2 USB
  ******************************************************************************/
-- 
2.47.2

